### PR TITLE
Feat: Add configuration in viml and through status function

### PR DESCRIPTION
**Pull request by shadmansaleh**
Addressing #113

Now lualine can be confugured through vimscript . Checkout docs for example.

### CODE REVIEW

**Comment by hoob3rt** [comment]
This looks like a nice solution. I will test this locally asap

**Comment by hoob3rt** [comment]
We should also change the full config example section to have component options. I actually have no idea how to do it in viml and can't test it

**Comment by hoob3rt** [comment]
also it would be better to have a variable to run lualine. Instead of calling `require'lualine`.status()` something like `lualine.enabled = true`

**Comment by shadmansaleh** [comment]
>We should also change the full config example section to have component options. I actually have no idea how to do it in viml and can't test it

Can you edit the lua example and add options that you want to add then I'll try to port it in viml.
We will probably have to do somthing like

```vim
'lualine_b':{'branch', {1: 'signify',color_added':'#.....',},},
```

>also it would be better to have a variable to run lualine. Instead of calling require'lualine'.status() something like lualine.enabled = true

I donn't think it's something that can be done while being 100% lua for now. Only file that get sourced automaticaly are vim files in plugin/ dir.
Without the require call none of lua code will be loaded. Also I think using a veriable to load a plugin is weired.

**Comment by hoob3rt** [comment]
I think I stumbled across a bug while doing the configuration. Can you confirm if this works properly?
```lua
lualine.sections.lualine_a = {{'filename', full_name = true} }
```

**Comment by hoob3rt** [comment]
> I donn't think it's something that can be done while being 100% lua for now. Only file that get sourced automaticaly are vim files in plugin/ dir.

True. I've tested this. I'm ditching this idea

**Comment by shadmansaleh** [comment]
>I think I stumbled across a bug while doing the configuration. Can you confirm if this works properly?

What beheviour are you seeing? (a bit more details(

**Comment by hoob3rt** [comment]
> > I think I stumbled across a bug while doing the configuration. Can you confirm if this works properly?
> 
> What beheviour are you seeing? (a bit more details)

Well it's simply not working for me. Not expanding the filename at all

**Comment by hoob3rt** [comment]
> > > I think I stumbled across a bug while doing the configuration. Can you confirm if this works properly?
> > 
> > 
> > What beheviour are you seeing? (a bit more details)
> 
> Well it's simply not working for me. Not expanding the filename at all

I actually noticed why it's not working. I will submit a fix to master and after that supply you the config

**Comment by hoob3rt** [comment]
Nevermind, the problem is that this branch is not up to date with master. Please resolve this first

**Comment by shadmansaleh** [comment]
>Nevermind, the problem is that this branch is not up to date with master. Please resolve this first

Sorry I couldn't understand .

**Comment by hoob3rt** [comment]
> > Nevermind, the problem is that this branch is not up to date with master. Please resolve this first
> 
> Sorry I couldn't understand .

merge master into this branch then we will be able to test this with proper configs

File: lua/lualine.lua [hunk 1]
```diff
@@ -35,8 +35,20 @@ M.inactive_sections = {
   lualine_z = {  }
 }
 
-M.extensions = {
-}
+M.extensions = { }
+
+function apply_viml_configuration()
```
Comment (hoob3rt) [inline lua/lualine.lua:40]: should be local

**Comment by hoob3rt** [comment]
This is the config I would like to have in readme both for lua and viml.
```lua
lualine.options = {
  theme = 'gruvbox',
  section_separators = {'', ''},
  component_separators = {'', ''},
  icons_enabled = true,
}
lualine.sections = {
  lualine_a = { {'mode', upper = true} },
  lualine_b = { {'branch', icon = ''} },
  lualine_c = { {'filename', file_status = true} },
  lualine_x = { 'encoding', 'fileformat', 'filetype' },
  lualine_y = { 'progress' },
  lualine_z = { 'location'  },
}
lualine.inactive_sections = {
  lualine_a = {  },
  lualine_b = {  },
  lualine_c = { 'filename' },
  lualine_x = { 'location' },
  lualine_y = {  },
  lualine_z = {   }
}
lualine.extensions = { 'fzf' }
```

This is basically a default config but it highlights how you can configure components

File: lua/lualine.lua (hunk 1 above)
Comment (shadmansaleh) [reply lua/lualine.lua:40]: Ofcource it should .

**Comment by shadmansaleh** [comment]
```lua
lualine.sections.lualine_a = {{'filename', full_name = true} }
```

I've found your bug . The options name was full_path not full_name lol.

**Comment by shadmansaleh** [comment]
Alright I've updated the docs and apply_viml_options function

I've chosen 'provider' as key to set component when dict is used to configure component_options. Unless you have problem with that name I think we're done with this pr.

The example config in viml ended up being

```vim
let g:lualine = {
    \'options' : {
    \  'theme' : 'gruvbox',
    \  'section_separators' : ['', ''],
    \  'component_separators' : ['', ''],
    \  'icons_enabled' : v:true,
    \},
    \'sections' : {
    \  'lualine_a' : [ {'provider': 'mode', 'upper': v:true,}, ],
    \  'lualine_b' : [ {'provider': 'branch', 'icon': '',}, ],
    \  'lualine_c' : [ {'provider': 'filename', 'file_status': v:true,}, ],
    \  'lualine_x' : [ 'encoding', 'fileformat', 'filetype' ],
    \  'lualine_y' : [ 'progress' ],
    \  'lualine_z' : [ 'location'  ],
    \},
    \'inactive_sections' : {
    \  'lualine_a' : [  ],
    \  'lualine_b' : [  ],
    \  'lualine_c' : [ 'filename' ],
    \  'lualine_x' : [ 'location' ],
    \  'lualine_y' : [  ],
    \  'lualine_z' : [  ],
    \},
    \'extensions' : [ 'fzf' ],
    \}
lua require("lualine").status()

```

**Comment by shadmansaleh** [comment]
I've added configuration through status function.
Now a configuration table can be directly given to status function and lualine will be configured acordingly. Configuring this way keeps all the lualine configuration in oneplace and it's easier to find. Also most lua plugins can be configured in this manner. I;ve changed the lua example to :

```lua
use {
  'hoob3rt/lualine.nvim',
  requires = {'kyazdani42/nvim-web-devicons', opt = true},
  config = function()
    require('lualine').status{
      options = {
        theme = 'gruvbox',
        section_separators = {'', ''},
        component_separators = {'', ''},
        icons_enabled = true,
      },
      sections = {
        lualine_a = { {'mode', upper = true} },
        lualine_b = { {'branch', icon = ''} },
        lualine_c = { {'filename', file_status = true} },
        lualine_x = { 'encoding', 'fileformat', 'filetype' },
        lualine_y = { 'progress' },
        lualine_z = { 'location'  },
      },
      inactive_sections = {
        lualine_a = {  },
        lualine_b = {  },
        lualine_c = { 'filename' },
        lualine_x = { 'location' },
        lualine_y = {  },
        lualine_z = {   }
      },
      extensions = { 'fzf' }
    }
  end
}
```

**Comment by hoob3rt** [comment]
I like what you did here. The `provider` name is okay. Wouldn't it be better to parse it this way though?
```vim
let g:lualine = {
    \'sections' : {
    \  'lualine_a' : [ ['mode', { 'upper': v:true }], ],
    \},
    \}
```

This will look closer to what is currently working on the lua side but will require us to have an additional function to parse vimscript config.

**Comment by hoob3rt** [comment]
Works great now! Thanks :)

**Pull request merged**